### PR TITLE
ZooKeeper Backend: Authnetication and Authorization support

### PR DESCRIPTION
**Pull request by eyal-lupu**
Provide a fix to issue #1058. 
With that fix users can set Zookeeper AddAuth string and ACLs. If none is provided the existing default (OPEN_ACL_UNSAFE) is kept. 

The PR also updates the relevant documentation.

### CODE REVIEW

File: physical/zookeeper.go [hunk 1]
```diff
@@ -52,16 +53,63 @@ func newZookeeperBackend(conf map[string]string) (Backend, error) {
...
+    var owner string
+    var schema string
+    var schemaAndOwner string
+    schemaAndOwner, ok = conf["znode_owner"]
+    if !ok {
+        owner = "anyone"
+        schema = "world"
+    } else {
+        parsedSchemaAndOwner := strings.SplitN(schemaAndOwner, ":", 2)
+        if !(len(parsedSchemaAndOwner)==2) {
```
Comment (jefferai) [inline physical/zookeeper.go]: Minor stylistic comment (here and below), for consistency with the rest of the Vault codebase: please use
```
if len(parsedSchemaAndOwner) != 2 {
```

File: physical/zookeeper.go [hunk 2]
```diff
@@ -52,16 +53,63 @@ func newZookeeperBackend(conf map[string]string) (Backend, error) {
...
+
+
 	// Attempt to create the ZK client
 	client, _, err := zk.Connect(strings.Split(machines, ","), time.Second)
 	if err != nil {
 		return nil, fmt.Errorf("client setup failed: %v", err)
 	}
 
+    // If auth_info provided - attempt to authenticate 
+    if owner != "" {
```
Comment (jefferai) [inline physical/zookeeper.go]: I believe this should check both the owner and schema for non-empty values, since schema is used the line below.

**Comment by jefferai** [comment]
A few minor comments, but looking good!

File: physical/zookeeper.go [hunk 3]
```diff
@@ -26,6 +26,7 @@ const (
 type ZookeeperBackend struct {
 	path   string
 	client *zk.Conn
+    acl     []zk.ACL
```
Comment (siddharthist) [inline physical/zookeeper.go]: Looks like a whitespace issue here. Did you run `gofmt`?

**Comment by eyal-lupu** [comment]
Updated the files as requested above.

**Comment by jefferai** [comment]
@siddharthist Any chance you are running ZK and want to give this a whirl for testing before I merge it?

File: website/source/docs/config/index.html.md [hunk 4]
```diff
@@ -246,6 +246,39 @@ For Zookeeper, the following options are supported:
...
 
+The following optional settings can be used to configure zNode ACLs
+  * `auth_info` (optional) - Authentication string to match Zookeeper AddAuth format
+      'schema:auth'. A simple example would be 'digest:UserName:Password' - to 
+      authenticate as UserName using Password
+  * `znode_owner` (optional) - Schema and User parts of Zookeeper ACL format. Expected format 
+      is 'schema:user-ACL-match. Vault will always set all permissions (CRWDA) to that ACL, some examples:
+    * 'digest:UserName:HIDfRvTv623G==' - The user 'UserName' with the 'HIDfRvTv623G==' (second colon is part
+      of the user information)
+    * 'ip:127.0.01' - Access from localhost only
```
Comment (siddharthist) [inline website/source/docs/config/index.html.md]: Looks like you might have meant 127.0.0.1?

**Comment by siddharthist** [comment]
@jefferai I'm not sure I could in a reasonable timeframe.

**Comment by jefferai** [comment]
@siddharthist OK. I will merge this then; if you get a chance to test later and report your findings, great!

**Pull request merged**